robot-server/robot_server/persistence/_migrations/up_to_3.py

```
@@ -173,3 +190,95 @@ def _migrate_analysis_table(
             protocol_id=row.protocol_id,
             analyzer_version=row.analyzer_version,
         )
+
+
+def _migrate_db_commands(
+    source_db_file: Path, dest_db_file: Path, run_ids: List[str]
+) -> None:
+    """Migrate the run commands stored in the database.
+
+    Because there are potentially tens or hundreds of thousands of commands in total,
+    this is the most computationally expensive part of the migration. We distribute
+    the work across subprocesses. Each subprocess extracts, migrates, and inserts
+    all of the commands for a single run.
+    """
+    # We'll use a lock to make sure only one process is accessing the database at once.
+    #
+    # Concurrent access would be safe in the sense that SQLite would always provide
+    # isolation. But, when there are conflicts, we'd have to deal with SQLite retrying
+    # transactions or raising SQLITE_BUSY. A Python-level lock is simpler and more
+    # reliable.
+    manager = multiprocessing.Manager()
+    lock = manager.Lock()
+
+    with multiprocessing.Pool(
+        # One worker per core of the OT-2's Raspberry Pi.
+        # We're compute-bound, so more workers would just thrash.
+        #
+        # Napkin math for the memory footprint:
+        # Suppose a very large protocol has ~10MB of commands (see e.g. RQA-443).
+        # The maximum number of runs at the time of writing is 20,
+        # so that's at most ~200MB total, which should be fine.
+        processes=4
+    ) as pool:
+        pool.map(
+            _migrate_db_commands_for_run,
+            ((source_db_file, dest_db_file, run_id, lock) for run_id in run_ids),
+        )
+
+
+def _migrate_db_commands_for_run(
```

Probably wrap this whole thing in a `try`/`except: pass`.